### PR TITLE
feat(sctpd): Enable mconfig and Sentry monitoring

### LINKED ISSUE #9509: Add Sentry to SCTPD
1. Remove sigmask of SEGV: https://github.com/magma/magma/blob/master/lte/gateway/c/sctpd/src/sctpd.cpp#L52 . Blocking SEGV will interfere with Sentry's ability to catch segfaults.
2. Copy how sentry is initialized in [sessiond_main.cpp](https://github.com/magma/magma/blob/master/lte/gateway/c/session_manager/sessiond_main.cpp#L210). You can ignore the bit about mconfig until https://github.com/magma/magma/issues/9508 is done. 
3. Modify CI to upload SCTPD debug artifact. See how this is done for SessionD in [.circleci/config.yml](https://github.com/magma/magma/blob/master/.circleci/config.yml#L354) (see also [this](https://github.com/magma/magma/blob/master/.circleci/config.yml#L358))

### LINKED ISSUE #10877: Enable support for mconfig in SCTPd
At the moment, SCTPd is not configurable from the cloud. The goal of this issue is to enable mconfig in SCTPd in order to configure a Sentry monitoring URL and sample rate as well as the log level of the service.

### PR DESCRIPTION
**Pull request by sebathomas**
## Summary

Introduce mconfig for SCTPd, and use it to set the log level and Sentry configuration. Enable Sentry if a URL is given.

This uses the default log level INFO, which is also used for the other services, instead of DEBUG as hardcoded before.

Closes #10877.

## Test Plan

Sent a message to Sentry test account via local SCTPd after configuring the Sentry URL in Swagger.

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/c/sctpd/src/BUILD.bazel
Comment (sebathomas) [inline lte/gateway/c/sctpd/src/BUILD.bazel:29]: This is so much more intuitive than the cmake file. :smile:

File: lte/gateway/c/sctpd/src/sctpd.cpp
Comment (sebathomas) [inline lte/gateway/c/sctpd/src/sctpd.cpp:82]: We used INFO because all other services seem to use INFO as default in the mconfig. Was there a reason to set DEBUG for this service?

File: lte/gateway/c/sctpd/src/sctpd.cpp
Comment (sebathomas) [inline lte/gateway/c/sctpd/src/sctpd.cpp]: It seems like the log levels between mconfig and verbosity are inverted so INFO is interpreted as ERROR (not just here but in other services too). Is this a known issue or am I misunderstanding something?

**Comment by sebathomas** [comment]
Looking into the CI errors...

File: lte/gateway/c/sctpd/src/sctpd.cpp
Comment (themarwhal) [reply lte/gateway/c/sctpd/src/sctpd.cpp]: That's interesting... How do other services handle it? I remember we have some logic to remap log values to glog log levels.

**Comment by themarwhal** [comment]
@sebathomas I think we may have to add "-ldl" to `bazel/external/sentry_native.BUILD`

**Comment by sebathomas** [comment]
> @sebathomas I think we may have to add "-ldl" to `bazel/external/sentry_native.BUILD`

That seems to fix it locally, thank you! Strange it wasn't needed for sessiond.

File: lte/gateway/c/sctpd/src/sctpd.cpp
Comment (themarwhal) [inline lte/gateway/c/sctpd/src/sctpd.cpp]: seeing this in CI
```
lte/gateway/c/sctpd/src/sctpd.cpp: In function 'int main()':
lte/gateway/c/sctpd/src/sctpd.cpp:101:10: warning: 'char* strncpy(char*, const char*, size_t)' specified bound 255 equals destination size [-Wstringop-truncation]
  101 |   strncpy(sentry_config.url_native,
      |   ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~
  102 |           mconfig.sentry_config().dsn_native().c_str(), MAX_URL_LENGTH);
```

File: lte/protos/mconfig/mconfigs.proto
Comment (themarwhal) [inline lte/protos/mconfig/mconfigs.proto]: Instead of taking this value here, let's take it from the Shared mconfig

**Comment by themarwhal** [comment]
> > @sebathomas I think we may have to add "-ldl" to `bazel/external/sentry_native.BUILD`
> 
> That seems to fix it locally, thank you! Strange it wasn't needed for sessiond.

yeah I'm not sure why exactly :/ maybe it was already linked in for something else

File: lte/gateway/c/sctpd/src/sctpd.cpp
Comment (themarwhal) [reply lte/gateway/c/sctpd/src/sctpd.cpp]: ah looks like same warning exists for sessiond? (would be great if you could address both 🙏 )

File: lte/protos/mconfig/mconfigs.proto
Comment (themarwhal) [reply lte/protos/mconfig/mconfigs.proto]: I think it's still useful to have a Magma SCTPD config section though. (We probably have other configurations that should be changeable from the cloud)  But let's pull sentry config from shared mconfig.

File: lte/gateway/c/sctpd/src/sctpd.cpp
Comment (sebathomas) [reply lte/gateway/c/sctpd/src/sctpd.cpp]: I think there's a mismatch between the values in `magma::orc8r::LogLevel` and the `magma_logging.h` constants. Putting this into the main
```cpp
magma::set_verbosity(magma::orc8r::LogLevel::DEBUG);
magma::set_verbosity(magma::orc8r::LogLevel::INFO);
magma::set_verbosity(magma::orc8r::LogLevel::WARNING);
magma::set_verbosity(magma::orc8r::LogLevel::ERROR);
magma::set_verbosity(magma::orc8r::LogLevel::FATAL);
magma::set_verbosity(MDEBUG);
magma::set_verbosity(MINFO);
magma::set_verbosity(MWARNING);
magma::set_verbosity(MERROR);
magma::set_verbosity(MFATAL);
```
results in:
```
I1216 15:31:06.167461 203384 magma_logging_init.h:28] Setting verbosity to 0
I1216 15:31:06.167547 203384 magma_logging_init.h:28] Setting verbosity to 1
I1216 15:31:06.167569 203384 magma_logging_init.h:28] Setting verbosity to 2
I1216 15:31:06.167587 203384 magma_logging_init.h:28] Setting verbosity to 3
I1216 15:31:06.167603 203384 magma_logging_init.h:28] Setting verbosity to 4
I1216 15:31:06.167619 203384 magma_logging_init.h:28] Setting verbosity to 4
I1216 15:31:06.167635 203384 magma_logging_init.h:28] Setting verbosity to 3
I1216 15:31:06.167675 203384 magma_logging_init.h:28] Setting verbosity to 2
I1216 15:31:06.167696 203384 magma_logging_init.h:28] Setting verbosity to 1
I1216 15:31:06.167714 203384 magma_logging_init.h:28] Setting verbosity to 0
```
So there's still a conversion missing. Sessiond, liagent and connection tracker have a function  `get_log_verbosity` that will also return the mconfig value directly if it's not overwritten by the config.

Comment (themarwhal) [reply lte/gateway/c/sctpd/src/sctpd.cpp]: yikes - we should probably fix that :D Can you file an issue?

Comment (sebathomas) [reply lte/gateway/c/sctpd/src/sctpd.cpp]: #10911

File: lte/gateway/c/sctpd/src/BUILD.bazel
Comment (electronjoe) [reply lte/gateway/c/sctpd/src/BUILD.bazel:29]: Yeah.. CMake exposes all sorts of the guts of building and linking as part of the configuration files - when that is absolutely not necessary and a better abstraction talks about e.g. Dependencies for the most part (which is the approach of Bazel / Buck).  This is precisely one of our arguments on Bazel - it is *hard* to contribute to our C/C++ today given the CMake situation (which admittedly is worse than it needs to be, if we were to completely fix all the CMake non-best-practices).  Bazel offers a minimum API and a "blessed path" primarily - so it's easy to replicate and do things the right way and you need not get into the guts of compilation.  /rant

File: lte/gateway/c/sctpd/src/sctpd.cpp
Comment (themarwhal) [inline lte/gateway/c/sctpd/src/sctpd.cpp]:
```suggestion
#include "orc8r/gateway/c/common/sentry/includes/SentryWrapper.h"
#include "orc8r/gateway/c/common/config/includes/MConfigLoader.h"
```
@electronjoe recently made it so that we fully specify include paths (https://github.com/magma/magma/pull/9926) We still have not made the switch for this in SessionD as it breaks CLion <-> CMake integration, apparently.

File: lte/gateway/c/sctpd/src/sctpd.cpp
Comment (sebathomas) [reply lte/gateway/c/sctpd/src/sctpd.cpp]: I added a manual null termination but I am not sure if there is a default way how to do it in Magma? I saw that many people recommend `strlcpy` but it doesn't seem to be used in the project.

File: orc8r/protos/BUILD.bazel
Comment (sebathomas) [inline orc8r/protos/BUILD.bazel:75]: I had to remove this to get the `#include` to work. It's now analogous to the mconfig settings in lte.

File: lte/protos/mconfig/mconfigs.proto
Comment (sebathomas) [reply lte/protos/mconfig/mconfigs.proto]: Done.

**Review by themarwhal — APPROVED** [review]

**Comment by sebathomas** [comment]
(Rebased and tried to fix sctpd.cpp include order warning.)

File: orc8r/protos/BUILD.bazel
Comment (themarwhal) [reply orc8r/protos/BUILD.bazel:75]: Yeah I wonder why I added this in the first place :o If the rest of C/C++ services still build fine, this is OK :)

File: lte/gateway/c/sctpd/src/sctpd.cpp
Comment (themarwhal) [inline lte/gateway/c/sctpd/src/sctpd.cpp]: Might be good to place this function in the common mconfig library so that it can be used by other services as well?

**Comment by themarwhal** [comment]
@sebathomas While we wait for reviews from @magma/approvers-agw-mme , want to add the debug artifact upload logic to https://github.com/magma/magma/blob/master/.github/workflows/build_all.yml#L222 ?

File: lte/gateway/c/sctpd/src/sctpd.cpp
Comment (sebathomas) [reply lte/gateway/c/sctpd/src/sctpd.cpp]: Done in #10969, will rebase and do it for sctpd when one of the commits is merged.

**Review by quentinDERORY — APPROVED** [review]

**Review by Neudrino — APPROVED** [review]

**Comment by ulaskozat** [comment]
@ssanadhya which line of code do we need to touch to make sure that sctpd is also upgraded?

**Comment by ssanadhya** [comment]
> @ssanadhya which line of code do we need to touch to make sure that sctpd is also upgraded?

@ulaskozat, @sebathomas, you can upgrade the Sctpd dependency by upgrading SCTPD_MIN_VERSION here https://github.com/magma/magma/blob/master/lte/gateway/release/build-magma.sh#L24

**Comment by ulaskozat** [comment]
> > @ssanadhya which line of code do we need to touch to make sure that sctpd is also upgraded?
> 
> @ulaskozat, @sebathomas, you can upgrade the Sctpd dependency by upgrading SCTPD_MIN_VERSION here https://github.com/magma/magma/blob/master/lte/gateway/release/build-magma.sh#L24

this PR does not break any backward compatibility as far as I see. The question is whether we should simply advise in 1.7 release to update sctpd service (as part of release notes) or make it mandatory with this line of code change.

File: lte/gateway/c/sctpd/src/sctpd.cpp
Comment (ulaskozat) [reply lte/gateway/c/sctpd/src/sctpd.cpp:82]: not really. INFO level is good.

File: lte/gateway/c/sctpd/src/sctpd.cpp
Comment (ulaskozat) [inline lte/gateway/c/sctpd/src/sctpd.cpp:90]: can we instead try to get it first from sctpd.yml file before hardwiring the default here?

Comment (ulaskozat) [reply lte/gateway/c/sctpd/src/sctpd.cpp:90]: in terms of priority yml file takes precedence over streamed down config which takes precedence over hardwired log level as the convention in AGW.

Comment (sebathomas) [reply lte/gateway/c/sctpd/src/sctpd.cpp:90]: There is no sctpd.yml config file yet. Should we introduce one for the log level?

Comment (sebathomas) [reply lte/gateway/c/sctpd/src/sctpd.cpp:90]: Added a second commit with a yaml config.

File: lte/gateway/c/sctpd/src/sctpd.cpp
Comment (sebathomas) [inline lte/gateway/c/sctpd/src/sctpd.cpp:120]: This conversion is used in a lot of services and could be extracted somewhere (maybe as a function in ServiceConfigLoader?).

File: lte/gateway/configs/sctpd.yml
Comment (ulaskozat) [inline lte/gateway/configs/sctpd.yml]: since this will take precedence over orc8r streamed configs, let's comment it out and place a comment on top stating that either it can be commented out here or an override config can be specified under `/var/opt/magma/configs/sctpd.yml`.

**Review by ulaskozat — APPROVED** [review]
LGTM except for the comment I provided about commenting out the line in yml file and placing a comment before it that explains that it overrides cloud config if commented out or yml file is specified in the override directory.

**Comment by ssanadhya** [comment]
> > > @ssanadhya which line of code do we need to touch to make sure that sctpd is also upgraded?
> > 
> > 
> > @ulaskozat, @sebathomas, you can upgrade the Sctpd dependency by upgrading SCTPD_MIN_VERSION here https://github.com/magma/magma/blob/master/lte/gateway/release/build-magma.sh#L24
> 
> this PR does not break any backward compatibility as far as I see. The question is whether we should simply advise in 1.7 release to update sctpd service (as part of release notes) or make it mandatory with this line of code change.

It is best to minimize upgrades to Sctpd (for hitless Magma package upgrades). We can just recommend upgrade to Sctpd in 1.7 release notes, without making it a mandatory dependency.

**Comment by sebathomas** [comment]
> LGTM except for the comment I provided about commenting out the line in yml file and placing a comment before it that explains that it overrides cloud config if commented out or yml file is specified in the override directory.

Good point, done. :+1:

**Comment by themarwhal** [comment]
@sebathomas Any clue why the Magma OAI Jenkins test is failing? (cc @rdefosse)

**Comment by rdefosse** [comment]
> @sebathomas Any clue why the Magma OAI Jenkins test is failing? (cc @rdefosse)

All 3 CI runs over the past 24 hours failed.

`https://jenkins-oai.eurecom.fr/job/MAGMA-MME-production/11084/artifact/test_results_magma_epc_rhel8.html`

Looks like MAGMA-MME SCTPD does not respond to eNB S1AP request.

Then `https://jenkins-oai.eurecom.fr/job/MAGMA-MME-production/11084/artifact/test_results_magma_oai_epc.html`

when it does, UE attachments are not easy..

I would suspect `sctpd` is incorrect.

**Comment by sebathomas** [comment]
> @sebathomas Any clue why the Magma OAI Jenkins test is failing? (cc @rdefosse)

Oops, I missed that, will take a look...

**Comment by sebathomas** [comment]
It seems that sctpd fails to start if no file gateway.mconfig can be found.

**Comment by rdefosse** [comment]
> It seems that sctpd fails to start if no file gateway.mconfig can be found.

Is there a template?

I guess if a template exists, it has to be added in the following dockerfiles:
`ci-scripts/docker/Dockerfile.mme.ci.rhel8  ci-scripts/docker/Dockerfile.mme.ci.ubuntu18  lte/gateway/docker/mme/Dockerfile.rhel8  lte/gateway/docker/mme/Dockerfile.ubuntu18.04`

**Comment by sebathomas** [comment]
I think I was wrong. The problem seems to be sctpd.yml not existing. If the mconfig file is missing there is an error message but the service doesn't crash.

**Comment by sebathomas** [comment]
Not sure what to do about it. It seems the other services also cannot handle it if the service.yml file is missing.

Is there something I need to do for the file to exist in the CI?

**Comment by rdefosse** [comment]
> 

Because in OAI deployment use case, we only activate MME and SCTPD services

**Comment by rdefosse** [comment]
Sorry I clicked on the wrong button

**Comment by sebathomas** [comment]
> > 
> 
> Because in OAI deployment use case, we only activate MME and SCTPD services

With the changes here, it would be necessary to create a file `/etc/magma/sctpd.yml` even for that use case.

**Comment by rdefosse** [comment]
@sebathomas 

OK. It's late here in France and I am in the midst of the quarterly SW release.

Let me have a look tomorrow morning. 

I will provide you with a patch to apply to your branch

**Comment by rdefosse** [comment]
@sebathomas 

Here is the patch I promised.

I've tested manually and the `sctpd` executable is now starting properly.

Apply the patch `git apply sctpd-mconfig-sentry.patch` and push again on your branch. CI will trigger again.

[sctpd-mconfig-sentry.zip](https://github.com/magma/magma/files/7895993/sctpd-mconfig-sentry.zip)

Regards
Raphael

**Comment by sebathomas** [comment]
@rdefosse Awesome, thank you. Just pushed again with the patch.

**Review by rdefosse — APPROVED** [review]

**Comment by sebathomas** [comment]
Rebasing on master.

**Pull request merged**
